File: kotlin/src/Currying.kt
package src

fun log(level: String, message: String) {
    println("[$level]: $message")
}

fun logForLevel(level: String): (String) -> Unit {
    return fun (message: String) {
        log(level, message)
    }
}

fun main() {
    val logDebug = logForLevel("DEBUG")
    logDebug("Hello, World!")  // [DEBUG] Hello, World!
}

File: rust/currying.rs
fn log(level: &str, message: &str) {
    println!("[{level}] {message}");
}

fn log_for_level(level: &str) -> impl Fn(&str) -> () + '_ {
    return move |message| log(level, message);
}

fn main() {
    let log_debug = log_for_level("DEBUG");
    log_debug("Hello, World!"); // [DEBUG] Hello, World!
}
